# Assignment 2: SQL

* In your data folder, there should have 4 csv files, `sh50.csv`, `sh50_oneyear_dailybar`, `stocks_info.csv`, `hundred_stocks_twoyears_daily_bar.csv`. 

* You may need to load below modules and connect to your own database.

In [1]:
import numpy as np
import pandas as pd
import sqlalchemy
import pymysql
from sqlalchemy import create_engine
%load_ext sql
pymysql.install_as_MySQLdb()

In [2]:
%sql mysql://1630018083:weihui1324@localhost/few1630018083?charset=utf8

'Connected: 1630018083@few1630018083'

* You may also need the sqlalchemy connector.

In [3]:
conn= create_engine("mysql+pymysql://{user}:{pw}@localhost/{db}?charset=utf8".format(user="1630018083", pw="weihui1324", db="HelloDB"))

In [4]:
conn_few= create_engine("mysql+pymysql://{user}:{pw}@localhost/{db}?charset=utf8".format(user="1630018083", pw="weihui1324", db="few1630018083"))

## Questions

1 Load data in above four csv files into four dataframes, named as `df_sh50`, `df_sh50_yearly`, `df_stocks_info`, `df_hundred_twoyears`.

In [5]:
df_sh50=pd.read_csv('./data/sh50.csv')
df_sh50_yearly=pd.read_csv('./data/sh50_oneyear_dailybar.csv')
df_stocks_info=pd.read_csv('./data/stocks_info.csv')
df_hundred_twoyears=pd.read_csv('./data/hundred_stocks_twoyears_daily_bar.csv')

2  Save data in above four csv files as four tables in MySQL database, using the same name as the csv files.

In [6]:
df_sh50.to_sql(con=conn_few, name='df_sh50', if_exists='replace', index=False)
df_sh50_yearly.to_sql(con=conn_few, name='df_sh50_yearly', if_exists='replace', index=False)
df_stocks_info.to_sql(con=conn_few, name='df_stocks_info', if_exists='replace', index=False)
df_hundred_twoyears.to_sql(con=conn_few, name='df_hundred_twoyears', if_exists='replace', index=False)

In [6]:
company_info=pd.read_sql('select * from company_info', conn_few)

3 Show tables in database.

In [48]:
%sql show tables;

7 rows affected.


[('company_info',),
 ('df_hundred_twoyears',),
 ('df_sh50',),
 ('df_sh50_yearly',),
 ('df_stocks_info',),
 ('sh50',),
 ('sh50_oneyear_dailybar',)]

**4 Show unique industries in company_info table.**

4.1 Use SQL.

In [25]:
%sql select DISTINCT industry from company_info

110 rows affected.


[('专用机械',),
 ('化工原料',),
 ('通信设备',),
 ('供气供热',),
 ('半导体',),
 ('证券',),
 ('航空',),
 ('元器件',),
 ('生物制药',),
 ('塑料',),
 ('白酒',),
 ('化学制药',),
 ('建筑施工',),
 ('饲料',),
 ('机械基件',),
 ('电气设备',),
 ('运输设备',),
 ('钢加工',),
 ('其他建材',),
 ('广告包装',),
 ('电器仪表',),
 ('乳制品',),
 ('铝',),
 ('服饰',),
 ('医疗保健',),
 ('软件服务',),
 ('保险',),
 ('综合类',),
 ('汽车配件',),
 ('空运',),
 ('轻工机械',),
 ('新型电力',),
 ('矿物制品',),
 ('染料涂料',),
 ('中成药',),
 ('农药化肥',),
 ('玻璃',),
 ('食品',),
 ('工程机械',),
 ('多元金融',),
 ('旅游服务',),
 ('互联网',),
 ('酒店餐饮',),
 ('小金属',),
 ('纺织机械',),
 ('超市连锁',),
 ('环境保护',),
 ('造纸',),
 ('家用电器',),
 ('橡胶',),
 ('全国地产',),
 ('百货',),
 ('影视音像',),
 ('水力发电',),
 ('区域地产',),
 ('商贸代理',),
 ('农业综合',),
 ('仓储物流',),
 ('纺织',),
 ('园区开发',),
 ('医药商业',),
 ('汽车整车',),
 ('种植业',),
 ('电脑设备',),
 ('林业',),
 ('旅游景点',),
 ('装修装饰',),
 ('日用化工',),
 ('公路',),
 ('渔业',),
 ('化纤',),
 ('水运',),
 ('特种钢',),
 ('水泥',),
 ('家居用品',),
 ('机场',),
 ('红黄药酒',),
 ('文教休闲',),
 ('啤酒',),
 ('电信运营',),
 ('房产服务',),
 ('出版业',),
 ('路桥',),
 ('摩托车',),
 ('铜',),
 ('化工机械',),
 ('汽车服务',),
 ('港口',),
 ('黄金',),
 ('批发业',),
 ('机床制造',),
 ('普钢',),
 ('其他商业',),
 ('软饮料',),
 ('农用机械',),
 ('水务',),
 ('铅锌',),
 ('煤炭开采',),
 ('银行',),
 ('陶瓷',),
 ('火力发电',),
 ('石油加工',),
 ('石油贸易',),
 ('石油开采',),
 ('焦炭加工',),
 ('船舶',),
 ('商品城',),
 ('公共交通',),
 ('铁路',),
 ('电器连锁',)]

4.2 Use Pandas Dataframe.

In [8]:
for i in company_info.name.unique():
    print (i)

N新余
N苏博特
N药石
N剑桥
贵州燃气
士兰微
华鑫股份
安达维尔
世运电路
德生科技
璞泰来
集泰股份
国农科技
联美控股
国立科技
老白干酒
华森制药
东宏股份
永福股份
通威股份
长盛轴承
中光防雷
韦尔股份
长江通信
合盛硅业
广哈通信
泰瑞机器
光库科技
英可瑞
金辰股份
江南嘉捷
华信新材
金银河
硕贝德
丹邦科技
振江股份
阿科力
凯伦股份
龙韵股份
海川智能
雅克科技
晶瑞股份
新易盛
中通国脉
华铭智能
鼎龙股份
庄园牧场
隆基股份
至纯科技
聚灿光电
太辰光
丽岛新材
佳力图
金一文化
天通股份
京运通
交大昂立
露笑科技
林洋能源
国电南瑞
新华保险
天宸股份
安科生物
浙江世宝
海立股份
春秋航空
天孚通信
雷迪克
光明乳业
吉大通信
科信技术
水晶光电
正泰电器
京山轻机
达威股份
华域汽车
纵横通信
博创科技
铭普光磁
江海股份
沃森生物
金风科技
北方华创
玉龙股份
太阳能
台海核电
依顿电子
中来股份
三超新材
飞凯材料
龙津药业
易德龙
恒生电子
华脉科技
三维通信
友讯达
跨境通
特变电工
利尔化学
东方日升
星网锐捷
蓝思科技
新元科技
中国巨石
西水股份
中国平安
超讯通信
惠威科技
东方电气
长电科技
福斯特
以岭药业
光威复材
桂发祥
三孚股份
三一重工
*ST大控
亿晶光电
中马传动
*ST东海A
光迅科技
郑煤机
邦讯技术
掌阅科技
*ST云网
中富通
寒锐钴业
腾龙股份
阳光电源
科达股份
容大感光
三花智控
五 粮 液
杰克股份
星帅尔
乐凯胶片
中际旭创
胜宏科技
湘电股份
正元智慧
中炬高新
中百集团
歌尔股份
嘉澳环保
酒鬼酒
耐威科技
思特奇
麦格米特
弘宇股份
视源股份
威孚高科
中国太保
中京电子
京华激光
风华高科
万盛股份
格力电器
锦富技术
金太阳
拓邦股份
深天马Ａ
英飞特
贵州茅台
中鼎股份
通宇通讯
通灵珠宝
武汉凡谷
中国武夷
福鞍股份
华东重机
汉商集团
三圣股份
得润电子
木林森
海量数据
奥飞娱乐
江丰电子
安靠智电
许继电气
祥源文化
创意信息
涪陵榨菜
西昌电力
杭萧钢构
红 宝 丽
广西广电
透景生命
大港股份
星源材质
泰格医药
森霸股份
保隆科技
久吾高科
北新建材
物产中大
*ST德力
长春高新
京泉华
浙江龙盛
浙江医药
金盾股份
中国人寿
路畅科技
顺发恒业
达刚

巴安水务
赣锋锂业
两面针
凤竹纺织
四维图新
双良节能
中牧股份
航天电子
汇通能源
美邦服饰
麦达数字
平治信息
粤高速Ａ
安彩高科
澄星股份
光大证券
阳光照明
明家联合
盛讯达
花王股份
大通燃气
中大力德
江苏阳光
杭州解百
蓝晓科技
长青集团
江苏国信
驰宏锌锗
鼎汉技术
必康股份
新光药业
华光股份
迪森股份
中色股份
拉夏贝尔
狮头股份
瑞茂通
众信旅游
中国交建
东兴证券
卫信康
中钢天源
香梨股份
吉电股份
天际股份
西部黄金
山鹰纸业
大金重工
天山股份
探路者
*ST昌九
栋梁新材
金海环境
尖峰集团
电连技术
光一科技
珠江实业
大秦铁路
浙江广厦
康力电梯
中航黑豹
龙源技术
亚邦股份
金鸿顺
上海梅林
东方能源
中闽能源
海航基础
同有科技
坚瑞沃能
扬农化工
益民集团
龙力生物
保税科技
北讯集团
惠达卫浴
苏常柴Ａ
皖通科技
上海洗霸
海南矿业
中国化学
赢时胜
同德化工
皇氏集团
漳泽电力
泰禾集团
中工国际
天域生态
漳州发展
石大胜华
古鳌科技
盛达矿业
晨化股份
三环集团
华贸物流
华电国际
兴源环境
景兴纸业
罗 牛 山
吉宏股份
深天地Ａ
焦点科技
银江股份
新天科技
富临精工
南通锻压
北京文化
天富能源
回天新材
吉林化纤
贵航股份
联泰环保
普路通
三 力 士
英特集团
北方导航
山东路桥
华控赛格
硅宝科技
理工环科
国检集团
贝瑞基因
四环生物
兆新股份
城投控股
兴齐眼药
森特股份
舒泰神
新黄浦
联明股份
维尔利
美亚柏科
*ST郑煤
新宏泽
燕塘乳业
中远海特
利欧股份
人人乐
华发股份
捷捷微电
金城医药
日照港
罗平锌电
中信银行
交通银行
惠天热电
网达软件
金健米业
易尚展示
红蜻蜓
利君股份
苏奥传感
福达股份
新潮能源
酒钢宏兴
长亮科技
乾景园林
华英农业
上海临港
西部矿业
大东方
华谊集团
中坚科技
四方冷链
世茂股份
大连重工
三星新材
大 东 南
鱼跃医疗
兴业证券
冠福股份
华铁科技
鸿博股份
同兴达
农发种业
山河智能
建研院
大连港
大康农业
招商轮船
申能股份
华宏科技
朗科科技
贵阳银行
融钰集团
中国核电
武汉中商
宁波建工
易见股份
中钨高新
外高桥
运达科技
新界泵业
瑞泰科技
东旭蓝天
联创互联
华宇软件
川大智胜
冰轮环境
上海钢联
美都能源
海泰发展
华联综超


**5 Show all the companies information under "软件服务" industry.**

5.1 Use SQL.

In [28]:
%sql select DISTINCT * from company_info where industry='软件服务'

160 rows affected.


[(59, 600406, '国电南瑞', '软件服务', '江苏', 43.39, 22.06, 24.29, 1754030.75, 1554292.0, 54272.51, 114886.0, 0.47, 0.354, 3.7, 5.54, 20031016, 473962.91, 1.95, 7.49, 12.11, 23.58, 12.02, 81465.0),
 (92, 600570, '恒生电子', '软件服务', '浙江', 90.55, 6.18, 6.18, 496096.06, 221085.0, 32502.06, 29003.58, 0.47, 0.469, 4.48, 12.64, 20031216, 133942.52, 2.17, 23.41, 41.15, 96.86, 17.56, 100025.0),
 (139, 300645, '正元智慧', '软件服务', '浙江', 1368.84, 0.17, 0.67, 79530.81, 67775.71, 681.71, 27784.84, 4.17, 0.029, 7.56, 6.9, 20170421, 14387.37, 2.16, 0.0, 0.0, 38.86, 0.83, 3236.0),
 (146, 300608, '思特奇', '软件服务', '北京', 348.03, 0.22, 0.88, 81293.43, 76198.17, 1929.95, 29530.21, 3.37, 0.08, 6.77, 5.46, 20170213, 18127.32, 2.07, -4.39, 239.94, 49.79, 2.01, 11494.0),
 (174, 603138, '海量数据', '软件服务', '北京', 76.2, 0.27, 1.07, 62705.81, 56070.82, 953.92, 13333.55, 1.25, 0.44, 3.46, 12.22, 20170306, 11683.0, 1.1, 0.0, 0.0, 21.51, 11.73, 15677.0),
 (180, 300366, '创意信息', '软件服务', '四川', 67.42, 2.09, 5.26, 374654.53, 220040.66, 8674.92, 191059.0, 3.64, 0.148, 5.44, 2.44, 20140127, 38458.4, 0.73, 72.12, 170.64, 25.27, 9.46, 15549.0),
 (222, 300508, '维宏股份', '软件服务', '上海', 57.8, 0.15, 0.57, 53064.4, 39024.57, 10444.65, 20013.46, 3.52, 1.217, 8.79, 10.67, 20160419, 21874.75, 3.85, 45.49, 95.91, 68.45, 45.54, 10964.0),
 (227, 2544, '杰赛科技', '软件服务', '广东', 259.26, 5.12, 5.16, 393695.28, 322186.16, 47700.21, 21579.09, 0.42, 0.057, 2.59, 7.58, 20110128, 53094.71, 1.03, 29.82, 12.81, 16.0, 1.36, 34914.0),
 (245, 2280, '联络互动', '软件服务', '浙江', 472.3, 12.65, 21.77, 1398823.0, 750404.88, 106541.57, 314105.16, 1.44, 0.015, 2.8, 3.44, 20090821, 67218.06, 0.31, 796.54, -87.31, 17.46, 0.43, 157044.0),
 (284, 300525, '博思软件', '软件服务', '福建', 212.06, 0.42, 0.72, 59848.9, 39545.62, 12300.4, 32046.21, 4.45, 0.18, 6.04, 8.14, 20160726, 15806.67, 2.19, 61.18, 1.61, 72.01, 9.54, 10635.0),
 (291, 2090, '金智科技', '软件服务', '江苏', 40.54, 2.07, 2.38, 404352.47, 296254.63, 60471.21, 26192.32, 1.1, 0.46, 4.76, 5.21, 20061208, 56567.01, 2.38, 59.11, 44.87, 24.06, 6.42, 17655.0),
 (407, 300559, '佳发安泰', '软件服务', '四川', 56.69, 0.4, 0.72, 70890.67, 52967.41, 1160.9, 28366.88, 3.95, 0.552, 8.49, 4.92, 20161101, 21949.07, 3.06, -5.83, -0.61, 55.69, 38.8, 11131.0),
 (414, 300096, '易联众', '软件服务', '福建', 0.0, 3.86, 4.3, 183600.69, 116812.76, 8729.0, 3277.05, 0.08, -0.122, 1.53, 9.9, 20100728, 17623.47, 0.41, 10.77, -442.04, 47.82, -16.85, 22445.0),
 (420, 600845, '宝信软件', '软件服务', '上海', 32.18, 5.24, 7.83, 710891.38, 519132.91, 75106.28, 136192.84, 1.74, 0.413, 5.58, 3.18, 19940311, 199943.09, 2.55, 22.97, 24.34, 28.3, 10.15, 39401.0),
 (421, 2401, '中远海科', '软件服务', '上海', 67.68, 3.01, 3.03, 141747.38, 124863.01, 9016.57, 8795.17, 0.29, 0.171, 2.64, 5.84, 20100506, 36189.37, 1.19, 1.21, 6.59, 18.95, 10.52, 22249.0),
 (439, 300065, '海兰信', '软件服务', '北京', 76.58, 2.74, 3.62, 214925.16, 147978.11, 7116.58, 66926.45, 1.85, 0.202, 3.31, 6.23, 20100326, 13832.7, 0.38, 10.93, 45.31, 47.24, 14.78, 21717.0),
 (454, 2065, '东华软件', '软件服务', '北京', 26.44, 27.95, 31.4, 1439059.13, 1172929.13, 31982.93, 234153.05, 0.75, 0.287, 2.91, 3.48, 20060823, 317053.63, 1.01, 17.71, 42.5, 33.4, 21.3, 97240.0),
 (476, 600455, '博通股份', '软件服务', '陕西', 195.56, 0.5, 0.62, 71287.27, 20086.0, 43532.45, 14566.34, 2.33, 0.161, 2.15, 19.47, 20040329, -8120.12, -1.3, -3.28, 325.19, 52.57, 8.7, 4674.0),
 (562, 300687, '赛意信息', '软件服务', '广东', 53.42, 0.2, 0.8, 83921.68, 80942.17, 291.2, 49710.67, 6.21, 1.1, 9.13, 6.46, 20170803, 14224.57, 1.78, 0.0, 0.0, 31.36, 13.34, 13585.0),
 (582, 300598, '诚迈科技', '软件服务', '江苏', 94.58, 0.2, 0.8, 50291.42, 41482.05, 697.02, 17531.67, 2.19, 0.31, 5.42, 6.9, 20170120, 16234.65, 2.03, -4.56, -27.72, 31.95, 6.5, 9562.0),
 (593, 662, '天夏智慧', '软件服务', '广西', 49.97, 2.88, 8.41, 584797.06, 165629.19, 198.16, 382785.72, 4.55, 0.248, 6.15, 2.69, 19961216, 43972.85, 0.52, -31.53, 52.58, 40.46, 46.47, 21238.0),
 (653, 300339, '润和软件', '软件服务', '江苏', 50.43, 5.0, 7.15, 521075.31, 176942.77, 60928.21, 217415.8, 3.04, 0.175, 5.0, 2.35, 20120718, 70266.15, 0.98, 20.81, 31.2,

5.2 Use Pandas Dataframe.

In [8]:
company_info[company_info.industry=="软件服务"]

Unnamed: 0  index    code   name industry area       pe  outstanding  \
59            59     59  600406   国电南瑞     软件服务   江苏    43.39        22.06   
92            92     92  600570   恒生电子     软件服务   浙江    90.55         6.18   
139          139    139  300645   正元智慧     软件服务   浙江  1368.84         0.17   
146          146    146  300608    思特奇     软件服务   北京   348.03         0.22   
174          174    174  603138   海量数据     软件服务   北京    76.20         0.27   
180          180    180  300366   创意信息     软件服务   四川    67.42         2.09   
222          222    222  300508   维宏股份     软件服务   上海    57.80         0.15   
227          227    227    2544   杰赛科技     软件服务   广东   259.26         5.12   
245          245    245    2280   联络互动     软件服务   浙江   472.30        12.65   
284          284    284  300525   博思软件     软件服务   福建   212.06         0.42   
291          291    291    2090   金智科技     软件服务   江苏    40.54         2.07   
407          407    407  300559   佳发安泰     软件服务   四川    56.69         0.40   
414          414    414  300096    易联众     软件服务   福建     0.00         3.86   
420          420    420  600845   宝信软件     软件服务   上海    32.18         5.24   
421          421    421    2401   中远海科     软件服务   上海    67.68         3.01   
439          439    439  300065    海兰信     软件服务   北京    76.58         2.74   
454          454    454    2065   东华软件     软件服务   北京    26.44        27.95   
476          476    476  600455   博通股份     软件服务   陕西   195.56         0.50   
562          562    562  300687   赛意信息     软件服务   广东    53.42         0.20   
582          582    582  300598   诚迈科技     软件服务   江苏    94.58         0.20   
593          593    593     662   天夏智慧     软件服务   广西    49.97         2.88   
653          653    653  300339   润和软件     软件服务   江苏    50.43         5.00   
691          691    691  300288   朗玛信息     软件服务   贵州    23.20         1.68   
704          704    704  600756   浪潮软件     软件服务   山东   167.76         3.24   
707          707    707  300033    同花顺     软件服务   浙江    48.75         2.64   
719          719    719  600556   ST慧球     软件服务   广西     0.00         3.95   
804          804    804    2153   石基信息     软件服务   北京    74.10         4.84   
807          807    807  300209   天泽信息     软件服务   江苏    75.31         2.44   
842          842    842  603383   顶点软件     软件服务   福建    89.32         0.21   
846          846    846  600797   浙大网新     软件服务   浙江    53.66         8.31   
...          ...    ...     ...    ...      ...  ...      ...          ...   
2907        2907   2907  300451   创业软件     软件服务   浙江    74.22         1.28   
2920        2920   2920  300386   飞天诚信     软件服务   北京    99.24         1.98   
2970        2970   2970  300496   中科创达     软件服务   北京   164.78         2.18   
2993        2993   2993  300479   神思电子     软件服务   山东   251.57         0.88   
2995        2995   2995  300324   旋极信息     软件服务   北京    45.55         5.96   
2998        2998   2998  300287    飞利信     软件服务   北京    27.50        10.41   
3011        3011   3011  600588   用友网络     软件服务   北京     0.00        14.60   
3047        3047   3047  300379    东方通     软件服务   北京    84.51         1.88   
3069        3069   3069  300532   今天国际     软件服务   深圳    51.80         0.66   
3070        3070   3070  603528   多伦科技     软件服务   江苏    60.30         1.88   
3085        3085   3085    2777   久远银海     软件服务   四川    99.61         0.60   
3088        3088   3088  300297   蓝盾股份     软件服务   广东    44.06         7.38   
3092        3092   3092    2230   科大讯飞     软件服务   安徽   359.98        11.78   
3094        3094   3094    2268  卫 士 通     软件服务   四川     0.00         5.51   
3103        3103   3103  300231   银信科技     软件服务   北京    33.26         2.67   
3108        3108   3108  300469   信息发展     软件服务   上海  1450.93         0.29   
3145        3145   3145  300036   超图软件     软件服务   北京    80.45         3.29   
3154        3154   3154  300350    华鹏飞     软件服务   深圳    48.07         2.19   
3177        3177   3177  300168   万达信息     软件服务   上海    92.52        10.09   
3187        3187   3187  300

**6 Show the company which has largest PE.**

6.1 Use SQL. 

In [31]:
%sql SELECT DISTINCT * FROM company_info ORDER BY pe DESC limit 1;

1 rows affected.


[(1445, 600733, 'S*ST前锋', '区域地产', '四川', 20163.08, 0.76, 1.98, 38522.1, 32295.6, 1480.91, 7322.18, 0.37, 0.002, 0.94, 53.53, 19960816, -8907.9, -0.45, 340.66, 105.91, 29.35, 1.67, 13520.0)]

6.2 Use Pandas Dataframe.

In [14]:
company_info[company_info.pe==np.max(company_info.pe)]

Unnamed: 0  index    code    name industry area        pe  outstanding  \
1445        1445   1445  600733  S*ST前锋     区域地产   四川  20163.08         0.76   

      totals  totalAssets   ...     bvps     pb  timeToMarket    undp  \
1445    1.98      38522.1   ...     0.94  53.53      19960816 -8907.9   

      perundp     rev  profit    gpr   npr  holders  
1445    -0.45  340.66  105.91  29.35  1.67  13520.0  

[1 rows x 25 columns]

**7 Show the number of companies in `stocks_info` table.**

7.1 Use SQL.

In [9]:
%sql select count(DISTINCT name) from df_stocks_info 

1 rows affected.


[(3445,)]

7.2 Use Pandas Dataframe.

In [24]:
df_stocks_info.name.count()

3445

**8 Show the entire information (e.g. area, industry,etc) of companies in sh50 index**

8.1 Use SQL.

In [49]:
%sql select * from company_info where code in (select code from df_sh50);

50 rows affected.


[(60, 601336, '新华保险', '保险', '北京', 31.25, 20.85, 31.2, 71477104.0, 0.0, 622100.0, 2395900.0, 7.68, 1.616, 20.39, 3.3, 20111216, 2789700.0, 8.94, -1.89, 5.35, 0.0, 4.36, 46900.0),
 (105, 601318, '中国平安', '保险', '深圳', 14.45, 108.33, 182.8, 616851584.0, 0.0, 3970000.0, 12283600.0, 6.72, 3.72, 24.52, 2.85, 20070301, 22220000.0, 12.16, 21.4, 17.36, 0.0, 10.0, 307286.0),
 (151, 601601, '中国太保', '保险', '上海', 27.63, 62.87, 90.62, 116817904.0, 0.0, 1285100.0, 6674200.0, 7.37, 1.206, 14.98, 2.96, 20071225, 4335500.0, 4.78, 21.42, 23.75, 0.0, 4.22, 77317.0),
 (162, 600519, '贵州茅台', '白酒', '贵州', 31.96, 12.56, 12.56, 12778004.0, 10480329.0, 1485118.38, 137496.44, 1.09, 15.908, 67.15, 10.1, 20010827, 7327905.5, 58.33, 59.4, 60.31, 89.93, 47.08, 63980.0),
 (201, 601628, '中国人寿', '保险', '北京', 26.89, 208.24, 282.65, 286852512.0, 0.0, 1882400.0, 5508100.0, 1.95, 0.94, 11.14, 3.05, 20070109, 14040899.0, 4.97, 19.98, 98.29, 0.0, 4.9, 156869.0),
 (218, 600887, '伊利股份', '乳制品', '内蒙', 28.49, 60.33, 60.79, 4713239.5, 2853463.25, 1293204.63, 277280.72, 0.46, 0.812, 3.98, 7.75, 19960312, 1358262.25, 2.23, 13.51, 12.47, 37.92, 9.47, 177933.0),
 (258, 600104, '上汽集团', '汽车整车', '上海', 11.54, 110.26, 116.83, 65232296.0, 35737896.0, 5098161.5, 5546412.5, 4.75, 2.12, 18.45, 1.76, 19971125, 10086611.0, 8.63, 14.23, 6.7, 13.19, 4.12, 104983.0),
 (375, 600029, '南方航空', '空运', '广东', 9.8, 70.23, 100.88, 21133400.0, 1749800.0, 15381299.0, 1511400.0, 1.5, 0.71, 5.02, 1.82, 20030725, 2331200.0, 2.31, 10.93, 9.46, 15.69, 7.33, 218751.0),
 (605, 601878, '浙商证券', '证券', '浙江', 56.6, 3.33, 33.33, 5144893.0, 0.0, 92830.53, 398078.09, 1.19, 0.26, 3.94, 4.66, 20170626, 404896.06, 1.21, 0.0, 0.0, 0.0, 24.71, 108890.0),
 (914, 600050, '中国联通', '电信运营', '上海', 142.59, 211.97, 302.34, 59056808.0, 9962275.0, 35827208.0, 2781473.75, 0.92, 0.06, 3.72, 2.22, 20021009, 3074480.25, 1.02, -0.66, 168.42, 25.71, 0.64, 787261.0),
 (945, 601881, '中国银河', '证券', '北京', 30.29, 6.0, 101.37, 24398524.0, 0.0, 26961.55, 2502739.0, 2.47, 0.34, 6.33, 2.13, 20170123, 1638712.75, 1.62, -8.4, -4.68, 0.0, 37.94, 123553.0),
 (961, 600000, '浦发银行', '银行', '上海', 6.64, 281.04, 293.52, 606383680.0, 0.0, 2527500.0, 8176000.0, 2.79, 1.45, 13.2, 0.96, 19991110, 10843900.0, 3.69, 3.18, 2.97, 0.0, 33.58, 192174.0),
 (978, 600019, '宝钢股份', '普钢', '上海', 11.85, 220.9, 221.03, 34661280.0, 13254327.0, 15759142.0, 4821146.0, 2.18, 0.528, 7.11, 1.17, 20001212, 5894787.5, 2.67, 43.12, 95.52, 10.89, 4.63, 462744.0),
 (979, 600030, '中信证券', '证券', '深圳', 21.08, 98.15, 121.17, 63177336.0, 0.0, 801862.5, 5437130.5, 4.49, 0.654, 12.13, 1.52, 20030106, 5087171.5, 4.2, 5.4, -0.18, 0.0, 27.81, 500513.0),
 (1063, 600016, '民生银行', '银行', '北京', 5.72, 295.52, 364.85, 571252480.0, 0.0, 4574800.0, 6474400.0, 1.77, 1.103, 9.89, 0.85, 20001219, 15708899.0, 4.31, -8.97, 2.76, 0.0, 37.99, 395946.0),
 (1389, 600340, '华夏幸福', '区域地产', '河北', 10.83, 29.55, 29.55, 33648380.0, 31025324.0, 279688.56, 465047.19, 1.57, 2.02, 9.12, 3.34, 20031230, 1819656.38, 6.16, 21.21, 35.3, 56.43, 20.16, 128168.0),
 (1436, 600958, '东方证券', '证券', '上海', 28.55, 31.26, 62.15, 22343832.0, 0.0, 44633.6, 1802878.25, 2.9, 0.423, 6.71, 2.4, 20150323, 876558.38, 1.41, 48.05, 40.15, 0.0, 37.24, 167822.0),
 (1497, 600036, '招商银行', '银行', '深圳', 8.71, 206.29, 252.2, 616923904.0, 0.0, 4799700.0, 6752300.0, 2.68, 2.332, 17.37, 1.56, 20020409, 23905100.0, 9.48, 3.38, 12.78, 0.0, 35.49, 218770.0),
 (1685, 600999, '招商证券', '证券', '深圳', 23.44, 49.04, 66.99, 25777122.0, 0.0, 53111.98, 2753394.0, 4.11, 0.56, 11.69, 1.66, 20091117, 1726285.75, 2.58, 11.21, 4.85, 0.0, 43.8, 100167.0),
 (1909, 601688, '华泰证券', '证券', '江苏', 23.13, 54.44, 71.63, 37462496.0, 0.0, 368614.84, 4583776.0, 6.4, 0.657, 12.0, 1.69, 20100226, 1732089.75, 2.42, 4.6, -2.13, 0.0, 36.84, 162144.0),
 (2005, 600518, '康美药业', '中成药', '广东', 26.56, 44.03, 49.47, 6533670.5, 5427972.5, 584594.5, 1136969.0, 2.3, 0.637, 6.28, 3.59, 20010319, 1038471.13, 2.1, 18.59, 21.58, 30.42, 16.15, 110288.0),
 (2014, 601169, '北京银行', '银行', '北京', 6.42, 182.48, 182.48, 227522592.0,

8.2 Use Pandas Dataframe.

In [27]:
mask=np.array([np.any(ele == df_sh50.name) for ele in company_info.name])
company_info[mask]

Unnamed: 0  index    code  name industry area      pe  outstanding  \
60            60     60  601336  新华保险       保险   北京   31.25        20.85   
105          105    105  601318  中国平安       保险   深圳   14.45       108.33   
151          151    151  601601  中国太保       保险   上海   27.63        62.87   
162          162    162  600519  贵州茅台       白酒   贵州   31.96        12.56   
201          201    201  601628  中国人寿       保险   北京   26.89       208.24   
218          218    218  600887  伊利股份      乳制品   内蒙   28.49        60.33   
258          258    258  600104  上汽集团     汽车整车   上海   11.54       110.26   
375          375    375  600029  南方航空       空运   广东    9.80        70.23   
605          605    605  601878  浙商证券       证券   浙江   56.60         3.33   
914          914    914  600050  中国联通     电信运营   上海  142.59       211.97   
945          945    945  601881  中国银河       证券   北京   30.29         6.00   
961          961    961  600000  浦发银行       银行   上海    6.64       281.04   
978          978    978  600019  宝钢股份       普钢   上海   11.85       220.90   
979          979    979  600030  中信证券       证券   深圳   21.08        98.15   
1063        1063   1063  600016  民生银行       银行   北京    5.72       295.52   
1389        1389   1389  600340  华夏幸福     区域地产   河北   10.83        29.55   
1436        1436   1436  600958  东方证券       证券   上海   28.55        31.26   
1497        1497   1497  600036  招商银行       银行   深圳    8.71       206.29   
1685        1685   1685  600999  招商证券       证券   深圳   23.44        49.04   
1909        1909   1909  601688  华泰证券       证券   江苏   23.13        54.44   
2005        2005   2005  600518  康美药业      中成药   广东   26.56        44.03   
2014        2014   2014  601169  北京银行       银行   北京    6.42       182.48   
2015        2015   2015  600837  海通证券       证券   上海   20.41        80.92   
2039        2039   2039  601211  国泰君安       证券   上海   18.84        47.33   
2090        2090   2090  601818  光大银行       银行   北京    5.51       398.10   
2093        2093   2093  600919  江苏银行       银行   江苏    7.49        57.95   
2146        2146   2146  600111  北方稀土      小金属   内蒙  111.69        36.33   
2183        2183   2183  601166  兴业银行       银行   福建    5.60       190.52   
2296        2296   2296  601229  上海银行       银行   上海    8.63         7.81   
2325        2325   2325  600547  山东黄金       黄金   山东   46.69        14.52   
2479        2479   2479  601800  中国交建     建筑施工   北京   15.30       117.47   
2498        2498   2498  601006  大秦铁路       铁路   山西    8.77       148.67   
2585        2585   2585  601328  交通银行       银行   上海    6.25       392.51   
2626        2626   2626  601985  中国核电     新型电力   北京   21.96        42.30   
2845        2845   2845  601988  中国银行       银行   北京    5.87      2107.66   
2880        2880   2880  600309  万华化学     化工原料   山东    9.64        25.95   
2906        2906   2906  601288  农业银行       银行   北京    5.53      2940.55   
2987        2987   2987  601857  中国石油     石油开采   北京   66.33      1619.22   
3013        3013   3013  600606  绿地控股     全国地产   上海   10.43        32.23   
3086        3086   3086  603993  洛阳钼业      小金属   河南   68.25       129.54   
3106        3106   3106  600048  保利地产     全国地产   广东   11.23       117.36   
3164        3164   3164  601989  中国重工       船舶   北京   84.29       183.62   
3172        3172   3172  601669  中国电建     建筑施工   北京   15.07        96.00   
3185        3185   3185  601398  工商银行       银行   北京    6.82      2696.12   
3188        3188   3188  601088  中国神华     煤炭开采   北京    9.04       164.91   
3256        3256   3256  600028  中国石化     石油加工   北京   14.41       955.58   
3297        3297   3297  601668  中国建筑     建筑施工   北京    8.25       297.32   
3333        3333   3333  601390  中国中铁     建筑施工   北京   13.66       183.28   
3336        3336   3336  601186  中国铁建     建筑施工   北京   11.47       115.03   
3350        3350   3350  601766  中国中车     运输设备   北京   34.49       229.18   

       totals   totalAssets    ...      bvps     pb  timeToMarket  \
60      31.20  7.147710e+07    ...     20.39   3.30      201

**9 Show the number of listed companies in each province**

9.1 Use SQL.

In [59]:
%sql select count(*) AS number, area from company_info GROUP BY area;

32 rows affected.


[(274, '上海'),
 (34, '云南'),
 (25, '内蒙'),
 (304, '北京'),
 (42, '吉林'),
 (114, '四川'),
 (49, '天津'),
 (13, '宁夏'),
 (102, '安徽'),
 (194, '山东'),
 (38, '山西'),
 (293, '广东'),
 (36, '广西'),
 (51, '新疆'),
 (379, '江苏'),
 (38, '江西'),
 (55, '河北'),
 (77, '河南'),
 (411, '浙江'),
 (30, '海南'),
 (266, '深圳'),
 (96, '湖北'),
 (100, '湖南'),
 (33, '甘肃'),
 (131, '福建'),
 (15, '西藏'),
 (27, '贵州'),
 (74, '辽宁'),
 (49, '重庆'),
 (47, '陕西'),
 (12, '青海'),
 (36, '黑龙江')]

9.2 Use Pandas Dataframe.

In [32]:
company_info['area'].value_counts()

浙江     411
江苏     379
北京     304
广东     293
上海     274
深圳     266
山东     194
福建     131
四川     114
安徽     102
湖南     100
湖北      96
河南      77
辽宁      74
河北      55
新疆      51
天津      49
重庆      49
陕西      47
吉林      42
山西      38
江西      38
广西      36
黑龙江     36
云南      34
甘肃      33
海南      30
贵州      27
内蒙      25
西藏      15
宁夏      13
青海      12
Name: area, dtype: int64

**10 Show the rank of provinces according to the number of SSH50 companies within the province**

10.1 Use SQL.

In [60]:
%sql select count(*) AS number, area from company_info where code in (select code from df_sh50) GROUP BY area order by number DESC;

13 rows affected.


[(20, '北京'),
 (11, '上海'),
 (4, '深圳'),
 (3, '广东'),
 (2, '江苏'),
 (2, '内蒙'),
 (2, '山东'),
 (1, '贵州'),
 (1, '浙江'),
 (1, '河北'),
 (1, '福建'),
 (1, '山西'),
 (1, '河南')]

10.2 Use Pandas dataframe.

In [40]:
mask=np.array([np.any(ele == df_sh50.name) for ele in company_info.name])
company_info[mask]['area'].value_counts()

北京    20
上海    11
深圳     4
广东     3
江苏     2
内蒙     2
山东     2
河南     1
山西     1
浙江     1
贵州     1
河北     1
福建     1
Name: area, dtype: int64

**11 Show the SSH50 companies in 广东 province**

11.1 Use SQL.

In [62]:
%sql SELECT * FROM company_info INNER JOIN df_sh50 ON company_info.code = df_sh50.code where area='广东';

3 rows affected.


[(375, 600029, '南方航空', '空运', '广东', 9.8, 70.23, 100.88, 21133400.0, 1749800.0, 15381299.0, 1511400.0, 1.5, 0.71, 5.02, 1.82, 20030725, 2331200.0, 2.31, 10.93, 9.46, 15.69, 7.33, 218751.0, 600029, '南方航空', 0.67),
 (2005, 600518, '康美药业', '中成药', '广东', 26.56, 44.03, 49.47, 6533670.5, 5427972.5, 584594.5, 1136969.0, 2.3, 0.637, 6.28, 3.59, 20010319, 1038471.13, 2.1, 18.59, 21.58, 30.42, 16.15, 110288.0, 600518, '康美药业', 1.16),
 (3106, 600048, '保利地产', '全国地产', '广东', 11.23, 117.36, 118.58, 63262392.0, 58543904.0, 341367.5, 1500900.25, 1.27, 0.697, 7.94, 1.31, 20060731, 6425729.5, 5.42, -13.31, 16.31, 32.21, 10.95, 188114.0, 600048, '保利地产', 2.29)]

11.2 Use Pandas dataframe.

In [41]:
mask=np.array([np.any(ele == df_sh50.name) for ele in company_info.name])
company_info[mask].groupby('area').get_group('广东')

Unnamed: 0  index    code  name industry area     pe  outstanding  \
375          375    375  600029  南方航空       空运   广东   9.80        70.23   
2005        2005   2005  600518  康美药业      中成药   广东  26.56        44.03   
3106        3106   3106  600048  保利地产     全国地产   广东  11.23       117.36   

      totals  totalAssets    ...     bvps    pb  timeToMarket        undp  \
375   100.88   21133400.0    ...     5.02  1.82      20030725  2331200.00   
2005   49.47    6533670.5    ...     6.28  3.59      20010319  1038471.13   
3106  118.58   63262392.0    ...     7.94  1.31      20060731  6425729.50   

      perundp    rev  profit    gpr    npr   holders  
375      2.31  10.93    9.46  15.69   7.33  218751.0  
2005     2.10  18.59   21.58  30.42  16.15  110288.0  
3106     5.42 -13.31   16.31  32.21  10.95  188114.0  

[3 rows x 25 columns]

**12.1 Set following constraints for the two tables.**


    For sh50:
     1. primary key(code)

    For stocks_info:
     1. primary key(code)
     2. totalassets >0
     
    For sh50_oneyears_dailybar:
     1. primary key(code, date)
     2. check (high > low)
     3. Foreign key (code) reference to stocks_info (code)     

In [8]:
%sql ALTER TABLE df_sh50 ADD PRIMARY KEY(code);

0 rows affected.


[]

In [9]:
%sql ALTER TABLE df_stocks_info ADD PRIMARY KEY(code);

0 rows affected.


[]

In [10]:
%sql alter table df_stocks_info add constraint totalAssets check (Id_P>0);

0 rows affected.


[]

In [8]:
%sql alter table df_sh50_yearly add CONSTRAINT fk foreign key (code) references df_sh50(code);

17633 rows affected.


[]

In [11]:
%sql select * from information_schema.TABLE_CONSTRAINTS t where CONSTRAINT_TYPE='FOREIGN KEY';

1 rows affected.


[('def', 'HelloDB', 'FK_bar', 'HelloDB', 'daily_bar', 'FOREIGN KEY')]

12.2 Suppose "海通证券" with code 600837 decides to delist from the exchange, please delete its information from sh50_oneyears_dailybar and stocks_info. Only complete this task by SQL.

In [10]:
%sql ALTER TABLE df_sh50_yearly drop foreign key fk;

0 rows affected.


[]

In [172]:
%sql DELETE FROM df_stocks_info WHERE code='600837';

1 rows affected.


[]

In [173]:
%sql DELETE FROM df_sh50_yearly WHERE code='600837';

365 rows affected.


[]

**13 Find the number of the companies with "科技" in its name and located in Guangdong province from stocks_info and calculate its average totalAssets.**

13.1 Use SQL.

In [154]:
%sql SELECT AVG(totalAssets) FROM df_stocks_info where area='广东' and name LIKE '%科技%';

1 rows affected.


[(360416.9017391304,)]

13.2 Use Pandas Dataframe.

In [87]:
mask=np.array([i[:2]=='科技' for i in df_stocks_info.groupby('area').get_group('广东').name])

In [85]:
mask2=np.array([i[-2:]=='科技' for i in df_stocks_info.groupby('area').get_group('广东').name])

In [94]:
df_stocks_info.groupby('area').get_group('广东')[mask | mask2].totalAssets.mean()

360416.9017391304

**14 Show the number of entries in the resulting table after four types of join (INNER, LEFT, RIGHT, FULL) on sh50 and stocks_info. **

14.1 Inner

In [163]:
%sql SELECT count(*) FROM df_sh50 INNER JOIN df_stocks_info ON df_sh50.name = df_stocks_info.name;

1 rows affected.


[(50,)]

14.2 Left

In [164]:
%sql SELECT count(*) FROM df_sh50 Left JOIN df_stocks_info ON df_sh50.name = df_stocks_info.name;

1 rows affected.


[(50,)]

14.3 Right

In [170]:
%sql SELECT count(*) FROM df_sh50 RIGHT JOIN df_stocks_info ON df_sh50.name = df_stocks_info.name;

1 rows affected.


[(3445,)]

14.4 Full

In [168]:
%sql SELECT name FROM df_sh50 UNION SELECT name FROM df_stocks_info;

3445 rows affected.


[('浦发银行',),
 ('民生银行',),
 ('宝钢股份',),
 ('中国石化',),
 ('南方航空',),
 ('中信证券',),
 ('招商银行',),
 ('保利地产',),
 ('中国联通',),
 ('上汽集团',),
 ('北方稀土',),
 ('万华化学',),
 ('华夏幸福',),
 ('康美药业',),
 ('贵州茅台',),
 ('山东黄金',),
 ('绿地控股',),
 ('海通证券',),
 ('伊利股份',),
 ('江苏银行',),
 ('东方证券',),
 ('招商证券',),
 ('大秦铁路',),
 ('中国神华',),
 ('兴业银行',),
 ('北京银行',),
 ('中国铁建',),
 ('国泰君安',),
 ('上海银行',),
 ('农业银行',),
 ('中国平安',),
 ('交通银行',),
 ('新华保险',),
 ('中国中铁',),
 ('工商银行',),
 ('中国太保',),
 ('中国人寿',),
 ('中国建筑',),
 ('中国电建',),
 ('华泰证券',),
 ('中国中车',),
 ('中国交建',),
 ('光大银行',),
 ('中国石油',),
 ('浙商证券',),
 ('中国银河',),
 ('中国核电',),
 ('中国银行',),
 ('中国重工',),
 ('洛阳钼业',),
 ('平安银行',),
 ('万 科Ａ',),
 ('国农科技',),
 ('世纪星源',),
 ('深振业Ａ',),
 ('全新好',),
 ('神州高铁',),
 ('中国宝安',),
 ('美丽生态',),
 ('深物业A',),
 ('南 玻Ａ',),
 ('沙河股份',),
 ('深康佳Ａ',),
 ('深中华A',),
 ('神州长城',),
 ('深深宝Ａ',),
 ('深华发Ａ',),
 ('深科技',),
 ('深赤湾Ａ',),
 ('深天地Ａ',),
 ('特 力Ａ',),
 ('飞亚达Ａ',),
 ('深圳能源',),
 ('国药一致',),
 ('深深房Ａ',),
 ('富奥股份',),
 ('中粮地产',),
 ('深桑达Ａ',),
 ('神州数码',),
 ('中国天楹',),
 ('华联控股',),
 ('深南电A',),
 ('深大通',),
 ('中集集团',),
 ('东旭蓝天',),
 ('中洲控股',),
 ('中航地产',),
 ('深纺织Ａ',),
 ('泛海控股',),
 ('康达尔',),
 ('德赛电池',),
 ('深天马Ａ',),
 ('方大集团',),
 ('皇庭国际',),
 ('深 赛 格',),
 ('华锦股份',),
 ('中金岭南',),
 ('农 产 品',),
 ('深圳华强',),
 ('中兴通讯',),
 ('北方国际',),
 ('中国长城',),
 ('华控赛格',),
 ('华侨城Ａ',),
 ('特发信息',),
 ('海王生物',),
 ('盐 田 港',),
 ('深圳机场',),
 ('天健集团',),
 ('广聚能源',),
 ('中信海直',),
 ('TCL 集团',),
 ('宜华健康',),
 ('中成股份',),
 ('丰原药业',),
 ('*ST川化',),
 ('华数传媒',),
 ('中联重科',),
 ('常山北明',),
 ('国际实业',),
 ('申万宏源',),
 ('东方市场',),
 ('美的集团',),
 ('潍柴动力',),
 ('许继电气',),
 ('冀东水泥',),
 ('金 融 街',),
 ('ST生化',),
 ('华意压缩',),
 ('胜利股份',),
 ('藏格控股',),
 ('山东地矿',),
 ('*ST沈机',),
 ('英特集团',),
 ('东旭光电',),
 ('渤海金控',),
 ('民生控股',),
 ('合肥百货',),
 ('小天鹅Ａ',),
 ('通程控股',),
 ('吉林化纤',),
 ('南京公用',),
 ('湖北宜化',),
 ('东阿阿胶',),
 ('徐工机械',),
 ('兴业矿业',),
 ('华天酒店',),
 ('粤高速Ａ',),
 ('张家界',),
 ('晨鸣纸业',),
 ('山东路桥',),
 ('鄂武商Ａ',),
 ('绿景控股',),
 ('海虹控股',),
 ('南华生物',),
 ('珠江控股',),
 ('中润资源',),
 ('珠海港',),
 ('华塑控股',),
 ('金路集团',),
 ('*ST烯碳',),
 ('丽珠集团',),
 ('渝 开 发',),
 ('国际医学',),
 ('荣安地产',),
 ('四环生物',),
 ('中兵红箭',),
 ('长航凤凰',),
 ('美菱电器',),
 ('广州浪奇',),
 ('岭南控股',),
 ('红 太 阳',),
 ('*ST紫学',),
 ('柳 工',),
 ('广弘控股',),
 ('大冷股份',),
 ('穗恒运Ａ',),
 ('华金资本',),
 ('万 家 乐',),
 ('万泽股份',),
 ('华映科技',),
 ('广宇发展',),
 ('云南白药',),
 ('粤电力Ａ',),
 ('中天金融',),
 ('佛山照明',),
 ('皖能电力',),
 ('中原环保',),
 ('金浦钛业',),
 ('金圆股份',),
 ('航天发展',),
 ('湖南投资',),
 ('江铃汽车',),
 ('创元科技',),
 ('靖远煤电',),
 ('沙隆达Ａ',),
 ('泰山石油',),
 ('神州信息',),
 ('西部创业',),
 ('莱茵体育',),
 ('万向钱潮',),
 ('昆百大Ａ',),
 ('烽火电子',),
 ('陕国投Ａ',),
 ('供销大集',),
 ('渝三峡Ａ',),
 ('海南海药',),
 ('海德股份',),
 ('泸州老窖',),
 ('苏常柴Ａ',),
 ('新大洲Ａ',),
 ('海马汽车',),
 ('粤宏远Ａ',),
 ('广东甘化',),
 ('威孚高科',),
 ('北部湾港',),
 ('哈工智能',),
 ('东北电气',),
 ('汇源通信',),
 ('金洲慈航',),
 ('黔轮胎Ａ',),
 ('启迪古汉',),
 ('太阳能',),
 ('平潭发展',),
 ('大通燃气',),
 ('*ST宝实',),
 ('古井贡酒',),
 ('东北制药',),
 ('兴蓉环境',),
 ('青岛双星',),
 ('建投能源',),
 ('韶能股份',),
 ('盛达矿业',),
 ('渤海股份',),
 ('神州易桥',),
 ('华媒控股',),
 ('阳光股份',),
 ('绵石投资',),
 ('西安旅游',),
 ('天首发展',),
 ('焦作万方',),
 ('*ST东海A',),
 ('京汉股份',),
 ('海航投资',),
 ('中油资本',),
 ('海螺型材',),
 ('新华联',),
 ('恒立实业',),
 ('吉林敖东',),
 ('长安汽车',),
 ('远大控股',),
 ('天茂集团',),
 ('高新发展',),
 ('*ST钒钛',),
 ('铜陵有色',),
 ('顺发恒业',),
 ('三木集团',),
 ('合金投资',),
 ('英 力 特',),
 ('风华高科',),
 ('茂化实华',),
 ('万方发展',),
 ('西王食品',),
 ('仁和药业',),
 ('格力电器',),
 ('泰达股份',),
 ('金岭矿业',),
 ('金科股份',),
 ('中钨高新',),
 ('*ST中富',),
 ('长春高新',),
 ('天夏智慧',),
 ('永安林业',),
 ('湖北广电',),
 ('经纬纺机',),
 ('美好置业',),
 ('荣丰控股',),
 ('金鸿控股',),
 ('盈方微',),
 ('阳 光 城',),
 ('上峰水泥',),
 ('当代东方',),
 ('智度股份',),
 ('恒天海龙',),
 ('襄阳轴承',),
 ('大连友谊',),
 ('山推股份',),
 ('视觉中国',),
 ('东方电子',),
 ('远兴能源',),
 ('中山公用',),
 ('东北证券',),
 ('华讯方舟',),
 ('建新矿业',),
 ('宝新能源',),
 ('ST亚太',),
 ('惠天热电',),
 ('*ST华泽',),
 ('滨海能源',),
 ('炼石有色',),
 ('沈阳化工',),
 ('模塑科技',),
 ('厦门信达',),
 ('正虹科技',),
 ('恒逸石化',),
 ('浙江震元',),
 ('双环科技',),
 ('大冶特钢',),
 ('河钢股份',),
 ('贝瑞基因',),
 ('京蓝科技',),
 ('锦龙股份',),
 ('丰乐种业',),
 ('中兴商业',),
 ('黑芝麻',),
 ('韶钢松山',),
 ('苏宁环球',),
 ('大地传媒',),
 ('新能泰山',),
 ('西安饮食',),
 ('湖南发展',),
 ('美锦能源',),
 ('京东方Ａ',),
 ('鲁 泰Ａ',),
 ('华东科技',),
 ('国元证券',),
 ('燕京啤酒',),
 ('四川美丰',),
 ('泰禾集团',),
 ('振华科技',),
 ('罗 牛 山',),
 ('中交地产',),
 ('南风